### PR TITLE
Normalize non-existent username/password to empty string in HTTPS

### PR DESCRIPTION
**Pull request by Keno**
The only situation in which I could imagine it being useful to distinguish
empty password from no password is being set is if the server requires a
username, but does not require a password. That combination seems to
be exceedingly rare, so unless some other usecase comes up, I won't
feel bad about the extra prompt.

Fixes certain clone operations over HTTPS.

### CODE REVIEW

File: base/libgit2/callbacks.jl [hunk 1]
```diff
@@ -144,20 +144,22 @@ function authenticate_userpass(creds::UserPasswordCredentials, libgit2credptr::P
 
     username = creds.user
     userpass = creds.pass
+    (username == nothing) && (username = "")
```
Comment (tkelman) [inline base/libgit2/callbacks.jl]: `===`

**Comment by tkelman** [comment]
dunno what's up with the appveyor timeouts, but it happened on both jobs so I restarted the build in case it's somehow related

**Comment by tkelman** [comment]
Oh. I see why it froze on AppVeyor now. The Pkg test is giving me a "Credentials required" popup now for "https://github.com" which isn't right.

**Comment by Keno** [comment]
Looks like we need to try with `username="", password=""` first then? For some reason I thought libgit2 did that automatically first. Oh well.

**Comment by tkelman** [comment]
The test is trying to pass credentials to avoid the prompt, which I think we should keep as working.

**Comment by Keno** [comment]
Ah, I see, the test is trying to pass `username="", password=""` explicitly, expecting them to be invalid. We can just change that to `nonexistentuserjusttestingwegetanerror/bar` or something like that

**Comment by Keno** [comment]
Ok, I added a way to explicitly specify whether one wants to prompt in the fact of incorrect credentials. I don't think there's any reasonable way to guess, so might as well be explicit.

File: base/libgit2/callbacks.jl [hunk 2]
```diff
@@ -28,18 +28,6 @@ function authenticate_ssh(creds::SSHCredentials, libgit2credptr::Ptr{Ptr{Void}},
         username_ptr, schema, host)
     isusedcreds = checkused!(creds)
 
-    errcls, errmsg = Error.last_error()
-    if errcls != Error.None
-        # Check if we used ssh-agent
-        if creds.usesshagent == "U"
-            println("ERROR: $errmsg ssh-agent")
```
Comment (tkelman) [inline base/libgit2/callbacks.jl]: this is technically a different issue, right? should it be a separate commit at least?

Comment (Keno) [inline base/libgit2/callbacks.jl]: Technically a separate issue, but is really worth backing this out of this pull request and making a separate commit?

Comment (tkelman) [inline base/libgit2/callbacks.jl]: it might have been worth separating into its own commit before you pushed, or at least mentioning in the commit message

**Comment by Keno** [comment]
Ok, I changed the approach here, in order to avoid the following:
```
julia> Pkg.clone("https://github.com/Keno/CS161")
INFO: Cloning CS161 from https://github.com/Keno/CS161
Username for 'https://github.com':Ken
Password for 'https://Ken@github.com':
Password for 'https://Ken@github.com':
Password for 'https://Ken@github.com':
Username for 'https://github.com':Keno
Password for 'https://Keno@github.com':
INFO: Computing changes...
```
Now, we never prompt for credentials unless we consider the credentials invalid (we tried three times and gave up), (note exceptions are ssh keys which we know have to exist for ssh credentials to work). Also, if through interactive prompting credentials get changed, they get reset to make sure that we do the retries (which I'm still not sure are necessary, but if we're doing them might as well be consistent about it). The default case is handled by passing in credentials which were manually invalidated so the immediately prompt. In my testing this has felt like the nicest way to handle all of this, but please report any usability issues.

Above example is now:
```
julia> Pkg.clone("https://github.com/Keno/CS161")
INFO: Cloning CS161 from https://github.com/Keno/CS161
Username for 'https://github.com':Ken
Password for 'https://Ken@github.com':
Username for 'https://github.com [Ken]':Keno
Password for 'https://Keno@github.com':
INFO: Computing changes...
```

File: base/libgit2/types.jl [hunk 3]
```diff
@@ -752,13 +747,21 @@ function checkused!(p::Union{UserPasswordCredentials, SSHCredentials})
...
-reset!(p::CachedCredentials) = foreach(reset!, values(p.cred))
+reset!(p::Union{UserPasswordCredentials, SSHCredentials}, cnt::Int=3) = (p.count = cnt; p)
+reset!(p::CachedCredentials) = (foreach(reset!, values(p.cred)); p)
 
 "Obtain the cached credentials for the given host+protocol (credid), or return and store the default if not found"
 get_creds!(collection::CachedCredentials, credid, default) = get!(collection.cred, credid, default)
 get_creds!(creds::AbstractCredentials, credid, default) = creds
 get_creds!(creds::Void, credid, default) = default
+function get_creds!(creds::Ref{Nullable{AbstractCredentials}}, credid, default)
+  if isnull(creds[])
```
Comment (tkelman) [inline base/libgit2/types.jl]: 4 space indent

File: base/libgit2/callbacks.jl [hunk 4]
```diff
@@ -213,56 +219,41 @@ function credentials_callback(libgit2credptr::Ptr{Ptr{Void}}, url_ptr::Cstring,
     # get credentials object from payload pointer
     creds = nothing
     creds_are_temp = true
```
Comment (tkelman) [inline base/libgit2/callbacks.jl]: no longer occurs anywhere else in the file? this had been triggering a `securezero!(creds)`, is that still needed?

Comment (Keno) [inline base/libgit2/callbacks.jl]: `securezero!` is triggered by finalizer. I don't think there's really anything else we can do and still retain usability and retries - at least until upstream gives us an argument that tells us whether or not the last call succeeded.
